cryptovec/src/platform/windows.rs

```diff
@@ -1,24 +1,99 @@
+use winapi::um::errhandlingapi::GetLastError;
 use winapi::um::memoryapi::{VirtualLock, VirtualUnlock};
+use winapi::um::sysinfoapi::{GetNativeSystemInfo, SYSTEM_INFO};
 
 use super::MemoryLockError;
 
+// To correctly lock/unlock memory, we need to know the pagesize:
+static PAGE_SIZE: OnceLock<usize> = OnceLock::new();
+// Store refcounters for all locked pages, since Windows doesn't handle that for us:
+static LOCKED_PAGES: Mutex<Vec<PageLock>> = Mutex::new(Vec::new());
+
+struct PageLock {
+    page_idx: usize,
+    lock_counter: usize,
+}
+
 /// Unlock memory on drop for Windows.
 pub fn munlock(ptr: *const u8, len: usize) -> Result<(), MemoryLockError> {
+    let page_indices = get_page_indices(ptr, len);
+    let mut locked_pages = LOCKED_PAGES
+        .lock()
+        .map_err(|e| MemoryLockError::new(format!("Accessing PageLocks failed: {e}")))?;
+    for page_idx in page_indices {
+        match locked_pages.binary_search_by_key(&page_idx, |pl| pl.page_idx) {
+            Ok(vec_idx) => {
+                #[allow(clippy::indexing_slicing)] // binary_search Ok() is always valid index
+                {
+                    locked_pages[vec_idx].lock_counter -= 1;
+                    if locked_pages[vec_idx].lock_counter == 0 {
+                        locked_pages.remove(vec_idx);
+                        unlock_page(page_idx)?;
+                    }
+                }
+            }
+            Err(_) => {
+                return Err(MemoryLockError::new(
+                    "Tried to unlock pointer from non-locked page!".into(),
+                ));
+            }
+        }
+    }
+    Ok(())
+}
+
+fn unlock_page(page_idx: usize) -> Result<(), MemoryLockError> {
     unsafe {
-        if VirtualUnlock(ptr as LPVOID, len as SIZE_T) == 0 {
-            return Err(MemoryLockError::new("VirtualUnlock".into()));
+        if VirtualUnlock((page_idx * get_page_size()) as LPVOID, 1 as SIZE_T) == 0 {
+            // codes can be looked up at https://learn.microsoft.com/en-us/windows/win32/debug/system-error-codes
+            let errorcode = GetLastError();
+            return Err(MemoryLockError::new(format!(
+                "VirtualUnlock: 0x{errorcode:x}"
+            )));
         }
     }
     Ok(())
 }
 
 pub fn mlock(ptr: *const u8, len: usize) -> Result<(), MemoryLockError> {
+    let page_indices = get_page_indices(ptr, len);
+    let mut locked_pages = LOCKED_PAGES
+        .lock()
+        .map_err(|e| MemoryLockError::new(format!("Accessing PageLocks failed: {e}")))?;
+    for page_idx in page_indices {
+        match locked_pages.binary_search_by_key(&page_idx, |pl| pl.page_idx) {
+            Ok(vec_idx) => {
+                #[allow(clippy::indexing_slicing)] // binary_search Ok() is always valid index
+                {
+                    locked_pages[vec_idx].lock_counter += 1;
+                }
+            }
+            Err(vec_idx) => {
+                lock_page(page_idx)?;
+                locked_pages.insert(
```

Comment:
Inserting into a `Vec` negates the savings from `binary_search_by_key` here. Something like a `BTreeMap` would be a better choice for storing counters.